src/Infrastructure.VS/GitWorkSpaceService.cs

```diff
@@ -0,0 +1,86 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2022 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.ComponentModel.Composition;
+using System.IO;
+using System.IO.Abstractions;
+using SonarLint.VisualStudio.Integration;
+
+namespace SonarLint.VisualStudio.Infrastructure.VS
+{
+    public interface IGitWorkspaceService
+    {
+        /// <summary>
+        /// Returns the current Git repository root
+        /// Or null if not a git repo
+        /// </summary>
+        string GetRepoRoot();
+    }
+
+    [Export(typeof(IGitWorkspaceService))]
+    internal class GitWorkSpaceService : IGitWorkspaceService
+    {
+        private readonly ILogger logger;
+        private readonly IWorkspaceService workspaceService;
+        private readonly IFileSystem fileSystem;
+
+        private const string gitFolder = ".git";
+
+        [ImportingConstructor]
+        public GitWorkSpaceService(IWorkspaceService workspaceService, ILogger logger) : this(workspaceService, logger, new FileSystem())
+        { }
+
+        internal GitWorkSpaceService(IWorkspaceService workspaceService, ILogger logger, IFileSystem fileSystem)
+        {
+            this.workspaceService = workspaceService;
+            this.logger = logger;
+            this.fileSystem = fileSystem;
+        }
+
+        public string GetRepoRoot()
+        {
+            var workspaceRoot = workspaceService.FindRootDirectory();
+            string gitRoot;
+
+            if(workspaceRoot == null)
+            {
+                return null;
+            }
+
+            var currentDir = new DirectoryInfo(workspaceRoot);
+
+            do
+            {
+                gitRoot = Path.Combine(currentDir.FullName, gitFolder);
+
+                if (fileSystem.Directory.Exists(gitRoot))
```

Comment:
Case-sensitivity: presumably (hopefully!) `Directory.Exists` is case-insensitive.

Unfortunately, we can't really test that our implementation is case-insensitive because we're using the file system abstraction in the tests, and we'd have to set up our mock to handle casing correctly, which wouldn't prove anything...

Reply:
If it weren't the case (pun not intended :D) I think most of the .net applications would have trouble with the file system.

Reply:
True; but we don't have any tests that prove whether our code is handling casing correctly. We do mock the `Directory.Exists` method; I guess just adding an assertion that it is called would be enough.